## Описание проекта
**Цель исследования:** проанализировать базу данных крупного сервиса для чтения книг по подписке.
**Инструменты:** сортировка, группировка, фильтрация, присоединение, подзапросы в FROM и WHERE.
**Библиотеки:** pandas, sqlalchemy

## Описание данных
В БД содержится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Таблица books**

Содержит данные о книгах:
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица authors**

Содержит данные об авторах:
- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица publishers**

Содержит данные об издательствах:
- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица ratings**

Содержит данные о пользовательских оценках книг:
- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица reviews**

Содержит данные о пользовательских обзорах:
- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя автора обзора;
- `text` — текст обзора.

## Задачи
1. Посчитайте, сколько книг вышло после 1 января 2000 года;
2. Для каждой книги посчитайте количество обзоров и среднюю оценку;
3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [1]:
#!/usr/bin/python
# -*- coding: utf-8 -*-

# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры подключения к БД
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных

# формируем строку соединения с БД.
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

for table in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    query = ''' SELECT * FROM {}; 
    '''.format(table) 

    request = pd.io.sql.read_sql(query, con = engine) 
    print(f'Таблица - "{table}"')
    display(request.head())
    print(request.info())
    print()

Таблица - "books"


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None

Таблица - "authors"


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None

Таблица - "publishers"


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None

Таблица - "ratings"


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None

Таблица - "reviews"


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None



**В целом данные готовы для обработки. Пропусков нет.**

## 1 Сколько книг вышло после 1 января 2000 года

In [2]:
query1 = ''' 
SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01'
'''

request1 = pd.io.sql.read_sql(query1, con = engine)


print(f'После 1 января 2000 года вышло {request1.iloc[0,0]} книг.')

После 1 января 2000 года вышло 819 книг.


## 2 Количество обзоров и средняя оценка для каждой книги

In [3]:
query2 = ''' 

SELECT 
    books.title AS book_name,
    books.book_id AS id,
    rev.n_rev AS n_reviews,
    rat.avg_rat AS avg_rating
FROM books LEFT JOIN 
                (SELECT
                    book_id,
                    AVG(rating) AS avg_rat
                FROM
                    ratings
                GROUP BY 
                    book_id
                ) AS rat ON rat.book_id = books.book_id
        LEFT JOIN
                (SELECT 
                    book_id,
                    COUNT(review_id) AS n_rev
                FROM
                    reviews
                GROUP BY 
                    book_id
                ) AS rev ON rev.book_id = books.book_id 
''' 

request2 = pd.io.sql.read_sql(query2, con = engine)
print('Количество обзоров(n_reviews) и средняя оценка(avg_rating) для каждой книги:')
request2

Количество обзоров(n_reviews) и средняя оценка(avg_rating) для каждой книги:


book_name   id  n_reviews  \
0            The Body in the Library (Miss Marple  #3)  652        2.0   
1                                            Galápagos  273        2.0   
2                             A Tree Grows in Brooklyn   51        5.0   
3    Undaunted Courage: The Pioneering First Missio...  951        2.0   
4                                          The Prophet  839        4.0   
..                                                 ...  ...        ...   
995                                Alice in Wonderland   64        4.0   
996          A Woman of Substance (Emma Harte Saga #1)   55        2.0   
997                                          Christine  148        3.0   
998  The Magicians' Guild (Black Magician Trilogy  #1)  790        2.0   
999                           The Plot Against America  828        2.0   

     avg_rating  
0      4.500000  
1      4.500000  
2      4.250000  
3      4.000000  
4      4.285714  
..          ...  
995    4.230769  
996    5.000000  
997    3.428571  
998    3.500000  
999    3.000000  

[1000 rows x 4 columns]

## 3 Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [4]:
query3 = ''' 
SELECT
    P.publisher_id,
    P.publisher,
    P.n_books
FROM
    (SELECT 
        books.publisher_id AS publisher_id,
        COUNT(books.book_id) AS n_books,
        publishers.publisher AS publisher 
    FROM 
        books 
        LEFT JOIN publishers ON books.publisher_id=publishers.publisher_id

    WHERE 
        num_pages>50
    GROUP BY
        books.publisher_id,
        publishers.publisher
    ORDER BY 
        n_books DESC) AS P
LIMIT 1 
''' 

request3 = pd.io.sql.read_sql(query3, con = engine)
print('Издательство, которое выпустило наибольшее число книг:')
request3

Издательство, которое выпустило наибольшее число книг:


publisher_id      publisher  n_books
0           212  Penguin Books       42

## 4 - Автор с самой высокой средней оценкой книг

Учитываются книги с 50 и более оценками

In [5]:
query4 = ''' 
SELECT 
    aut_avg.author,
    AVG(aut_avg.avg_rating) AS avg_avg_rating
FROM
        (SELECT 
            authors.author AS author, 
            AVG(ratings.rating) AS avg_rating
        FROM books
            LEFT JOIN authors ON books.author_id = authors.author_id
            LEFT JOIN ratings ON books.book_id = ratings.book_id
        GROUP BY 
             books.book_id, 
             authors.author_id
        HAVING
            COUNT(ratings.rating_id) >=50 
        ) AS aut_avg 
GROUP BY    
    aut_avg.author
ORDER BY
    avg_avg_rating DESC
LIMIT 1
 ''' 

request4 = pd.io.sql.read_sql(query4, con = engine)
print('Автор с самой высокой средней оценкой книг:')
request4

Автор с самой высокой средней оценкой книг:


author  avg_avg_rating
0  J.K. Rowling/Mary GrandPré        4.283844

## 5 - Среднее количество обзоров от пользователей

Учитываются пользователи, которые поставили больше 48 оценок

In [6]:
query5 = ''' 

SELECT
    AVG(rev.n_reviews)
FROM
    (SELECT
        username,
        COUNT(review_id) AS n_reviews
    FROM
        reviews
    WHERE 
        username IN (
                SELECT 
                    username
                FROM 
                    ratings
                GROUP BY 
                    username
                HAVING 
                    COUNT(rating_id) > 48
                    )
    GROUP BY 
        username
    ) AS rev
''' 

request5 = pd.io.sql.read_sql(query5, con = engine)
print(f'Среднее количество обзоров от пользователей {request5.iloc[0,0]}.')

Среднее количество обзоров от пользователей 24.0.


## ВЫВОДЫ:
- После 1 января 2000 года вышло 819 книг.
- В датасете `request2` представлено количество обзоров(n_reviews) и средняя оценка(avg_rating) для каждой книги.
- Издательство, которое выпустило наибольшее число книг - Penguin Books (42 книги).
- Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré (рейтинг 4.3).
- Среднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24.0 (то есть примерно каждая 2 оценка включает в себя отзыв книги).